File: Kattis/C++/Shortest-Path-1/main.cpp
#include <bits/stdc++.h>

using namespace std;
using i64 = long long;
const i64 INF = INT64_MAX / 4;
#define endl '\n'

vector<pair<i64, i64>> dij(i64 s, vector<vector<pair<i64, i64>>> &graph) {
  vector<pair<i64, i64>> ans(graph.size(), {INF, -1});
  priority_queue<pair<i64, i64>, vector<pair<i64, i64>>,
                 greater<pair<i64, i64>>>
      pq;

  ans[s] = {0, -1};
  pq.push({0, s});

  while (!pq.empty()) {
    auto crt = pq.top();
    pq.pop();
    i64 dist = crt.first;
    i64 node = crt.second;

    if (dist > ans[node].first) {
      continue;
    }

    for (auto &c : graph[node]) {
      i64 val = crt.first + c.second;
      if (val < ans[c.first].first) {
        ans[c.first] = {val, node};
        pq.push({val, c.first});
      }
    }
  }

  return ans;
}

int main() {
  i64 n, m, q, s;

  while (true) {
    cin >> n >> m >> q >> s;
    if (n == 0 and m == 0 and q == 0 and s == 0) {
      break;
    }

    vector<vector<pair<i64, i64>>> graph(n);

    for (i64 i = 0; i < m; i++) {
      i64 u, v, w;
      cin >> u >> v >> w;
      graph[u].push_back({v, w});
    }

    vector<pair<i64, i64>> dist = dij(s, graph);

    while (q--) {
      i64 eep;
      cin >> eep;

      if (dist[eep].first == INF) {
        cout << "IMPOSSIBLE" << endl;
      } else {
        cout << dist[eep].first << endl;
      }
    }
    cout << endl;
  }

  return 0;
}

File: Kattis/C++/Shortest-Path-1/main.py
from heapq import heappush, heappop


def dij(g, s):
    dist = [int(1e9) for i in range(len(g))]
    dist[s] = 0
    pq = []
    heappush(pq, (0, s))

    while (len(pq) > 0):
        d, u = heappop(pq)
        if (d > dist[u]):
            continue
        for v, w in g[u]:
            cand = dist[u] + w
            if cand < dist[v]:
                dist[v] = cand
                heappush(pq, (dist[v], v))

    return dist


def main():
    while True:
        n, m, q, s = map(int, input().split())

        if n == 0 and m == 0 and q == 0 and s == 0:
            break

        g = [[] for i in range(n)]
        for _ in range(m):
            u, v, w = map(int, input().split())
            g[u].append((v, w))

        ans = dij(g, s)

        for _ in range(q):
            t = int(input())
            if ans[t] != int(1e9):
                print(ans[t])
            else:
                print("Impossible")

        print()


main()
